File: src/optical_element/MATLAB/Furion_hole.m
classdef Furion_hole<handle
    %¾ØÐÎslits
    
    properties
        beam_out;%slitsÊä³ö¹âÊø
    end
    
    methods 
        function obj = Furion_hole(beam_in,center_x,center_y,r) 
            mask=(beam_in.X-center_x).^2+(beam_in.Y-center_y).^2<r^2;
            V=beam_in.field.*mask;
%             xx = beam_in.X(:);
%             yy = beam_in.Y(:);
%             vv = beam_in.field(:);
%             n = length(xx);
%             N = beam_in.N;
%        % coord =zeros(1,n);
%         for i = 1:n
%             if ~(sqrt((xx(i)-center_x).^2 + (yy(i)-center_y).^2)<r)
%                 vv(i) =0;
%             end
%         end
%        X=reshape(xx,[N,N]);
%        Y=reshape(yy,[N,N]);
%        V =reshape(vv,[N,N]);
        %k=find(coord);
        obj.beam_out = beam(beam_in.X,beam_in.Y,V,beam_in.wavelength);
        end
    end
    
end

File: src/optical_element/Furion_hole.cpp
#include "Furion_hole.h"

using namespace Furion_NS;

Furion_Hole::Furion_Hole()
{
    cout << "Furion_Hole ³õÊ¼»¯" << endl;
}

Furion_Hole::~Furion_Hole()
{

}

void Furion_Hole::Furion_hole(Beam* beam_in, std::vector<std::vector<double> >& center_x, std::vector<std::vector<double> >& center_y, double r, int N)
{
	double r2 = r * r;
	std::vector<std::vector<std::complex<double>> > V;
	for (int i = 0; i < N; i++)
	{
		for (int j = 0; j < N; j++)
		{
			if (std::pow((beam_in->X[i][j] - center_x[i][j]) * (beam_in->X[i][j] - center_x[i][j]) + (beam_in->Y[i][j] - center_y[i][j]) * (beam_in->Y[i][j] - center_y[i][j]), 2) < r2)
			{
				V[i][j] = beam_in->field[i][j];
			}
			else
			{
				V[i][j] = std::complex<double>(0, 0);
			}
		}
	}

	beam_out = new Beam(beam_in->X, beam_in->Y, V, beam_in->wavelength, beam_in->N);
}



